chain/client/src/client.rs

```diff
@@ -206,17 +243,40 @@ impl Handler<NetworkClientMessages> for ClientActor {
 
     fn handle(&mut self, msg: NetworkClientMessages, ctx: &mut Context<Self>) -> Self::Result {
         match msg {
-            NetworkClientMessages::Transaction(tx) => match self.validate_tx(tx) {
-                Ok(valid_transaction) => {
-                    self.tx_pool.insert_transaction(valid_transaction);
-                    NetworkClientResponses::ValidTx
+            NetworkClientMessages::Transaction(tx) => {
+                let runtime_adapter = self.runtime_adapter.clone();
+                let shard_id = runtime_adapter.account_id_to_shard_id(&tx.body.get_originator());
+                if let Ok(chunk_extra) = self.chain.get_latest_chunk_extra(shard_id) {
+                    match self.runtime_adapter.validate_tx(shard_id, chunk_extra.state_root, tx) {
+                        Ok(valid_transaction) => {
+                            debug!(
+                                "MOO recording a transaction. I'm {:?}, {}",
+                                self.block_producer.as_ref().unwrap().account_id,
+                                shard_id
+                            );
+                            self.shards_mgr.insert_transaction(shard_id, valid_transaction);
+                            NetworkClientResponses::ValidTx
+                        }
+                        Err(err) => NetworkClientResponses::InvalidTx(err),
+                    }
+                } else {
+                    NetworkClientResponses::NoResponse
```

Comment:
Is it expected that tx get dropped in this case?

Reply:
We should probably route it to a validator that is in the shard for this tx.
But if we are doing it already on networking level, here we can just drop - as we are assuming it doesn't arrive here unless it's some edge case (like we just switched the caring shards).